replica/table.cc

```diff
@@ -2517,6 +2523,25 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
     };
 }
 
+max_purgeable_fn table::get_max_purgeable_fn_for_cache_underlying_reader() const {
+    return [this](const dht::decorated_key& dk, ::is_shadowable is_shadowable) {
+        auto& sg = storage_group_for_token(dk.token());
+        auto max_purgeable_timestamp = api::max_timestamp;
+
+        // if a memtable with a minimum timestamp lower than the current maximum
+        // purgeable timestamp has the given key, the tombstone should not be purged
+        sg.for_each_active_memtable([&dk, is_shadowable, &max_purgeable_timestamp](const memtable& mt) {
+            // see get_max_purgeable_timestamp() in compaction.cc for comments on choosing min timestamp
+            api::timestamp_type memtable_min_timestamp = is_shadowable ? mt.get_min_live_row_marker_timestamp() : mt.get_min_live_timestamp();
+            if (memtable_min_timestamp <= max_purgeable_timestamp && mt.contains_partition(dk)) {
+                max_purgeable_timestamp = memtable_min_timestamp;
```

Comment:
not that it matters much in practice, but there's an off-by-one here, we can bypass contains_partition() check when min_timestamp == max_purgeable, since it won't have any effect on the current max_purgeable.

Reply:
Ack. Fixed it in v4.